File: src/main/java/com/kasisoft/libs/common/text/StringFBuilder.java
package com.kasisoft.libs.common.text;

import com.kasisoft.libs.common.utils.MiscFunctions;

import javax.validation.constraints.Min;
import javax.validation.constraints.NotNull;

import java.util.stream.IntStream;

import java.io.IOException;
import java.io.ObjectInputStream;
import java.io.ObjectOutputStream;
import java.io.Serializable;

import lombok.experimental.FieldDefaults;

import lombok.AccessLevel;

/**
 * StringF(ormatting)Builder  equivalent which supports formatting. This builder also supports negative indices which 
 * means that the original index is calculated beginning from the end of the buffer.
 * 
 * @author daniel.kasmeroglu@kasisoft.net
 */
@FieldDefaults(level = AccessLevel.PRIVATE)
public class StringFBuilder implements Serializable, StringLike<StringFBuilder> {

  private static final long serialVersionUID = 1050795857819832439L;
  
  // the original implementation
  StringBuilder  origin;
  
  /**
   * @see StringBuilder#StringBuilder()
   */
  public StringFBuilder() {
    origin = new StringBuilder();
  }

  /**
   * @see StringBuilder#StringBuilder(int)
   */
  public StringFBuilder(@Min(1) int capacity) {
    origin = new StringBuilder(capacity);
  }

  /**
   * @see StringBuilder#StringBuilder(CharSequence)
   */
  public StringFBuilder(@NotNull CharSequence seq) {
    origin = new StringBuilder(seq);
  }

  @Override
  public @Min(0) int length() {
    return origin.length();
  }

  @Override
  public @Min(1) int capacity() {
    return origin.capacity();
  }

  @Override
  public StringFBuilder ensureCapacity(@Min(1) int minimum) {
    origin.ensureCapacity(minimum);
    return this;
  }

  @Override
  public StringFBuilder trimToSize() {
    origin.trimToSize();
    return this;
  }

  @Override
  public StringFBuilder setLength(@Min(0) int newlength) {
    origin.setLength(newlength);
    return this;
  }

  @Override
  public char charAt(int index) {
    return origin.charAt(adjustIndex(index, false));
  }

  @Override
  public int codePointAt(int index) {
    return origin.codePointAt(adjustIndex(index, false));
  }

  @Override
  public int codePointBefore(int index) {
    return origin.codePointBefore(adjustIndex(index, false));
  }

  @Override
  public int codePointCount(int begin, int end) {
    return origin.codePointCount(adjustIndex(begin, false), adjustIndex(end, true));
  }

  @Override
  public int offsetByCodePoints(int index, int codepointoffset) {
    return origin.offsetByCodePoints(adjustIndex(index, false), codepointoffset);
  }

  @Override
  public StringFBuilder getChars(int start, int end, @NotNull char[] destination, int destbegin) {
    origin.getChars(adjustIndex(start, false), adjustIndex(end, true), destination, MiscFunctions.adjustIndex(destination.length, destbegin, false));
    return this;
  }

  @Override
  public StringFBuilder setCharAt(int index, char ch) {
    origin.setCharAt(adjustIndex(index, false), ch);
    return this;
  }

  @Override
  public StringFBuilder setCodepointAt(int index, int codepoint) {
    index     = adjustIndex(index, false);
    var count = Character.charCount(codepoint);
    delete(index, index + count);
    return this;
  }

  @Override
  public @NotNull StringFBuilder append(@NotNull Object obj) {
    origin.append(obj);
    return this;
  }

  @Override
  public @NotNull StringFBuilder append(@NotNull CharSequence sequence) {
    origin.append(sequence);
    return this;
  }

  @Override
  public @NotNull StringFBuilder append(@NotNull CharSequence sequence, int start, int end) {
    origin.append(sequence, MiscFunctions.adjustIndex(sequence.length(), start, false), MiscFunctions.adjustIndex(sequence.length(), end, true));
    return this;
  }

  @Override
  public @NotNull StringFBuilder append(@NotNull char[] charray) {
    origin.append(charray);
    return this;
  }

  @Override
  public @NotNull StringFBuilder append(@NotNull char[] charray, int offset, int length) {
    origin.append(charray, MiscFunctions.adjustIndex(charray.length, offset, false), length);
    return this;
  }
  
  @Override
  public @NotNull StringFBuilder append(boolean value) {
    origin.append(value);
    return this;
  }

  @Override
  public @NotNull StringFBuilder append(char value) {
    origin.append(value);
    return this;
  }

  @Override
  public @NotNull StringFBuilder append(int value) {
    origin.append(value);
    return this;
  }

  @Override
  public @NotNull StringFBuilder appendCodePoint(int codepoint) {
    origin.appendCodePoint(codepoint);
    return this;
  }

  @Override
  public @NotNull StringFBuilder append(long value) {
    origin.append(value);
    return this;
  }

  @Override
  public @NotNull StringFBuilder append(float value) {
    origin.append(value);
    return this;
  }

  @Override
  public @NotNull StringFBuilder append(double value) {
    origin.append(value);
    return this;
  }

  @Override
  public @NotNull StringFBuilder delete(int start, int end) {
    origin.delete(adjustIndex(start, false), adjustIndex(end, true));
    return this;
  }

  @Override
  public @NotNull StringFBuilder deleteCharAt(int index) {
    origin.deleteCharAt(adjustIndex(index, false));
    return this;
  }

  @Override
  public @NotNull StringFBuilder replace(int start, int end, @NotNull String str) {
    origin.replace(adjustIndex(start, false), adjustIndex(end, true), str);
    return this;
  }

  @Override
  public @NotNull String substring(int start) {
    return origin.substring(adjustIndex(start, false));
  }

  @Override
  public @NotNull CharSequence subSequence(int start, int end) {
    return origin.subSequence(adjustIndex(start, false), adjustIndex(end, true));
  }

  @Override
  public @NotNull String substring(int start, int end) {
    return origin.substring(adjustIndex(start, false), adjustIndex(end, true));
  }

  @Override
  public @NotNull StringFBuilder insert(int index, @NotNull char[] charray, int offset, int length) {
    origin.insert(adjustIndex(index, false), charray, MiscFunctions.adjustIndex(charray.length, offset, false), length);
    return this;
  }

  @Override
  public @NotNull StringFBuilder insert(int offset, @NotNull Object obj) {
    origin.insert(adjustIndex(offset, false), obj);
    return this;
  }

  @Override
  public @NotNull StringFBuilder insert(int offset, @NotNull char[] value) {
    origin.insert(adjustIndex(offset, false), value);
    return this;
  }

  @Override
  public @NotNull StringFBuilder insert(int offset, @NotNull CharSequence value) {
    origin.insert(adjustIndex(offset, false), value);
    return this;
  }

  @Override
  public @NotNull StringFBuilder insert(int offset, @NotNull CharSequence value, int start, int end) {
    origin.insert(adjustIndex(offset, false), value, MiscFunctions.adjustIndex(value.length(), start, false), MiscFunctions.adjustIndex(value.length(), end, true));
    return this;
  }

  @Override
  public @NotNull StringFBuilder insert(int offset, boolean value) {
    origin.insert(adjustIndex(offset, false), value);
    return this;
  }

  @Override
  public @NotNull  StringFBuilder insert(int offset, char value) {
    origin.insert(adjustIndex(offset, false), value);
    return this;
  }

  @Override
  public @NotNull StringFBuilder insert(int offset, int value) {
    origin.insert(adjustIndex(offset, false), value);
    return this;
  }

  @Override
  public @NotNull StringFBuilder insert(int offset, long value) {
    origin.insert(adjustIndex(offset, false), value);
    return this;
  }

  @Override
  public @NotNull StringFBuilder insert(int offset, float value) {
    origin.insert(adjustIndex(offset, false), value);
    return this;
  }

  @Override
  public @NotNull StringFBuilder insert(int offset, double value) {
    origin.insert(adjustIndex(offset, false), value);
    return this;
  }

  @Override
  public int indexOf(@NotNull String str) {
    return origin.indexOf(str);
  }

  @Override
  public int indexOf(@NotNull String str, int index) {
    return origin.indexOf(str, adjustIndex(index, false));
  }

  @Override
  public int lastIndexOf(@NotNull String str) {
    return origin.lastIndexOf(str);
  }

  @Override
  public int lastIndexOf(@NotNull String str, int index) {
    return origin.lastIndexOf(str, adjustIndex(index, false));
  }
  
  @Override
  public @NotNull StringFBuilder reverse() {
    origin.reverse();
    return this;
  }

  @Override
  public @NotNull String toString() {
    return origin.toString();
  }
  
  private int adjustIndex(int index, boolean isEnd) {
    return MiscFunctions.adjustIndex(origin.length(), index, isEnd);
  }

  @Override
  public @NotNull IntStream chars() {
    return origin.chars();
  }
  
  @Override
  public @NotNull IntStream codePoints() {
    return origin.codePoints();
  }
  
  @Override
  public int compareTo(@NotNull StringFBuilder another) {
    if (this == another) {
      return 0;
    }
    return origin.compareTo(another.origin);
  }
  
  private void writeObject(ObjectOutputStream s) throws IOException {
    s.writeObject(origin);
  }

  private void readObject(ObjectInputStream s) throws IOException, ClassNotFoundException {
    origin = (StringBuilder) s.readObject();
  }

} /* ENDCLASS */

File: src/main/kotlin/com/kasisoft/libs/common/text/StringFBuilder.kt
package com.kasisoft.libs.common.text

import com.kasisoft.libs.common.utils.*

import javax.validation.constraints.*

import java.util.stream.*

import java.io.*

/**
 * StringF(ormatting)Builder  equivalent which supports formatting. This builder also supports negative indices which
 * means that the original index is calculated beginning from the end of the buffer.
 *
 * @author daniel.kasmeroglu@kasisoft.com
 */
open class StringFBuilder: Serializable, StringLike<StringFBuilder> {

    companion object {

        private const val serialVersionUID = 1050795857819832439L

    }

    // the original implementation
    private var origin: StringBuilder

    override val length: Int
        get() = origin.length

    /**
     * @see StringBuilder.StringBuilder
     */
    constructor() {
        origin = StringBuilder()
    }

    /**
     * @see StringBuilder.StringBuilder
     */
    constructor(capacity: @Min(1) Int) {
        origin = StringBuilder(capacity)
    }

    /**
     * @see StringBuilder.StringBuilder
     */
    constructor(seq: CharSequence) {
        origin = StringBuilder(seq)
    }

    override fun get(index: Int): Char = origin[adjustIndex(index)]

    override fun capacity(): Int = origin.capacity()

    override fun ensureCapacity(minimum: Int): StringFBuilder {
        origin.ensureCapacity(minimum)
        return this
    }

    override fun trimToSize(): StringFBuilder {
        origin.trimToSize()
        return this
    }

    override fun setLength(newlength: Int): StringFBuilder {
        origin.setLength(newlength)
        return this
    }

    override fun codePointAt(index: Int): Int = origin.codePointAt(adjustIndex(index))

    override fun codePointBefore(index: Int): Int = origin.codePointBefore(adjustIndex(index))

    override fun codePointCount(begin: Int, end: Int): Int = origin.codePointCount(adjustIndex(begin), adjustIndex(end, true))

    override fun offsetByCodePoints(index: Int, codepointoffset: Int): Int = origin.offsetByCodePoints(adjustIndex(index), codepointoffset)

    override fun getChars(start: Int, end: Int, destination: CharArray, destbegin: Int): StringFBuilder {
        origin.getChars(
            adjustIndex(start),
            adjustIndex(end, true),
            destination,
            adjustIndex(destbegin, false, destination.size)
        )
        return this
    }

    override fun setCharAt(index: Int, ch: Char): StringFBuilder {
        origin.setCharAt(adjustIndex(index), ch)
        return this
    }

    override fun setCodepointAt(index: Int, codepoint: Int): StringFBuilder {
        val idx   = adjustIndex(index)
        val count = Character.charCount(codepoint)
        delete(idx, idx + count)
        return this
    }

    override fun append(obj: Any): StringFBuilder {
        origin.append(obj)
        return this
    }

    override fun append(sequence: CharSequence): StringFBuilder {
        origin.append(sequence)
        return this
    }

    override fun append(sequence: CharSequence, start: Int, end: Int): StringFBuilder {
        origin.append(
            sequence,
            adjustIndex(start, false, sequence.length),
            adjustIndex(end, true, sequence.length)
        )
        return this
    }

    override fun append(charray: CharArray): StringFBuilder {
        origin.append(charray)
        return this
    }

    override fun append(charray: CharArray, offset: Int, length: Int): StringFBuilder {
        origin.append(charray, adjustIndex(offset, false, charray.size), length)
        return this
    }

    override fun append(value: Boolean): StringFBuilder {
        origin.append(value)
        return this
    }

    override fun append(value: Char): StringFBuilder {
        origin.append(value)
        return this
    }

    override fun append(value: Int): StringFBuilder {
        origin.append(value)
        return this
    }

    override fun appendCodePoint(codepoint: Int): StringFBuilder {
        origin.appendCodePoint(codepoint)
        return this
    }

    override fun append(value: Long): StringFBuilder {
        origin.append(value)
        return this
    }

    override fun append(value: Float): StringFBuilder {
        origin.append(value)
        return this
    }

    override fun append(value: Double): StringFBuilder {
        origin.append(value)
        return this
    }

    override fun delete(start: Int, end: Int): StringFBuilder {
        origin.delete(adjustIndex(start), adjustIndex(end, true))
        return this
    }

    override fun deleteCharAt(index: Int): StringFBuilder {
        origin.deleteCharAt(adjustIndex(index))
        return this
    }

    override fun replace(start: Int, end: Int, str: String): StringFBuilder {
        origin.replace(adjustIndex(start), adjustIndex(end, true), str)
        return this
    }

    override fun substring(start: Int): String = origin.substring(adjustIndex(start))

    override fun subSequence(startIndex: Int, endIndex: Int): CharSequence = origin.subSequence(adjustIndex(startIndex), adjustIndex(endIndex, true))

    override fun substring(start: Int, end: Int): String = origin.substring(adjustIndex(start), adjustIndex(end, true))

    override fun insert(index: Int, charray: CharArray, offset: Int, length: Int): StringFBuilder {
        origin.insert(adjustIndex(index), charray, adjustIndex(offset, false, charray.size), length)
        return this
    }

    override fun insert(offset: Int, obj: Any): StringFBuilder {
        origin.insert(adjustIndex(offset), obj)
        return this
    }

    override fun insert(offset: Int, value: CharArray): StringFBuilder {
        origin.insert(adjustIndex(offset), value)
        return this
    }

    override fun insert(offset: Int, value: CharSequence): StringFBuilder {
        origin.insert(adjustIndex(offset), value)
        return this
    }

    override fun insert(offset: Int, value: CharSequence, start: Int, end: Int): StringFBuilder {
        origin.insert(
            adjustIndex(offset),
            value,
            adjustIndex(start, false, value.length),
            adjustIndex(end, true, value.length)
        )
        return this
    }

    override fun insert(offset: Int, value: Boolean): StringFBuilder {
        origin.insert(adjustIndex(offset), value)
        return this
    }

    override fun insert(offset: Int, value: Char): StringFBuilder {
        origin.insert(adjustIndex(offset), value)
        return this
    }

    override fun insert(offset: Int, value: Int): StringFBuilder {
        origin.insert(adjustIndex(offset), value)
        return this
    }

    override fun insert(offset: Int, value: Long): StringFBuilder {
        origin.insert(adjustIndex(offset), value)
        return this
    }

    override fun insert(offset: Int, value: Float): StringFBuilder {
        origin.insert(adjustIndex(offset), value)
        return this
    }

    override fun insert(offset: Int, value: Double): StringFBuilder {
        origin.insert(adjustIndex(offset), value)
        return this
    }

    override fun indexOf(str: String): Int = origin.indexOf(str)

    override fun indexOf(str: String, index: Int): Int = origin.indexOf(str, adjustIndex(index))

    override fun lastIndexOf(str: String): Int = origin.lastIndexOf(str)

    override fun lastIndexOf(str: String, index: Int): Int = origin.lastIndexOf(str, adjustIndex(index))

    override fun reverse(): StringFBuilder {
        origin.reverse()
        return this
    }

    override fun toString(): String = origin.toString()

    override fun chars(): IntStream = origin.chars()

    override fun codePoints(): IntStream = origin.codePoints()

    override fun compareTo(other: StringFBuilder): Int =
        if (this === other) {
            0
        } else {
            origin.compareTo(other.origin)
        }

    @Throws(IOException::class)
    private fun writeObject(s: ObjectOutputStream) {
        s.writeObject(origin)
    }

    @Throws(IOException::class, ClassNotFoundException::class)
    private fun readObject(s: ObjectInputStream) {
        origin = s.readObject() as StringBuilder
    }

    private fun adjustIndex(index: Int, isEnd: Boolean = false, length: Int = origin.length): Int =
        if (index < 0) {
            length + index
        } else if ((index == 0) && isEnd) {
            length
        } else {
            index
        }

} /* ENDCLASS */
